vllm/core/scheduler.py

```diff
@@ -510,6 +529,67 @@ def __init__(
         # for processing and deallocation by the free_finished_seq_groups()
         self._async_stopped: List[SequenceGroup] = []
 
+    def _fetch_kv_thread(self):
+        assert self.kv_cache_shared_dict is not None
+
+        def hash_list(input):
+            import hashlib
+
+            import numpy as np
+            input_bytes = np.array(input).tobytes()
+            hash_object = hashlib.blake2b(input_bytes)
+            hash_hex = hash_object.hexdigest()
+            return int(hash_hex[:16], 16)
+
+        def get_kv_and_hidden_states(prefix):
+            kv_cache, hidden_states = get_kv_transfer_group(
+            ).recv_kv_caches_and_hidden_states_cpu(prefix)
+
+            return prefix, kv_cache, hidden_states
+
+        def put_to_shared_dict(prefix, kv_cache, hidden_states):
+            # Store the kv_cache and hidden_states in the shared dict.
+            # TODO: need to check whether have memory copy!!!
+            start_time_stamp = time.time()
+            kv_cache_hpu = kv_cache.to("hpu", non_blocking=True)
+            hidden_states_hpu = hidden_states.to("hpu", non_blocking=True)
+            end_time_stamp = time.time()
+            self.kv_cache_shared_dict.add_item(
+                prefix, [kv_cache_hpu, hidden_states_hpu])
+            if torch.distributed.get_rank() == 0:
+                logger.info(
+                    "putting kv cache to shared dict, takes: %3f seconds"
+                    "start time: %s, end time: %s",
+                    end_time_stamp - start_time_stamp, start_time_stamp,
+                    end_time_stamp)
+
+        while True:
+            if self.fetching_thread_should_shutdown:
+                logger.info("The fetching thread is shutting down.")
+                return
+            if not self.fetching_kv.empty():
+                self.scheduler_profiler.start('internal', 'fetching_kv')
+                seq_group = self.fetching_kv.get()
+                hash_prefix = hash_list(seq_group.prompt_token_ids)
+                prefix, kv_cache, hidden_states = get_kv_and_hidden_states(
+                    hash_prefix)
+                put_to_shared_dict(prefix, kv_cache, hidden_states)
+                if seq_group is not None:
+                    self.waiting.append(self.fetching_kv.get())
```

Comment:
what if seq_group is none? in this case, fetching_kv was put twice, but only get once.

Reply:
I think this should never be none... I will take a test.